text/0000-fn_name_macro.md

```diff
@@ -0,0 +1,61 @@
+- Feature Name: fn_name_macro
+- Start Date: 2019-11-14
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+<!-- https://github.com/rust-lang/rfcs/issues/1743 -->
+
+# Summary
+[summary]: #summary
+
+This RFC adds an additional macro, `function!`, to the `core` crate. When invoked, the macro expands to the name of the function that contains the call site.
+
+# Motivation
+[motivation]: #motivation
+
+This is a useful extension of Rust's existing debug reporting: `file!`, `line!`, `column!` and `module_path!`.
+
+For most people the name of the function is a much more immediate way to understand the context of a message than file name and line number.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For debug information about what's happening within the program there's several useful macros that you might use. One of them is the `function!` macro, which expands to the name of the current function. If used outside of a function it causes a compilation error.
```

Comment:
I think returning an `Option<&'static str>` or some special name like `"(top-level)"` is more useful than a compilation error. It is surprising when some macro cannot be used outside a function.

Reply:
Definitely not an Option, that would make it much more painful to use.

Reply:
Well if we generalize this to items rather than to functions then there is always an enclosing item wherever this is used except directly in top level modules so erroring becomes much less common.